File: data/harnesses/Plus/cpp/secure/CWE-125_07/main.cpp
#include <vector>
#include <string>
#include <stdexcept>
#include <iostream>
#include <functional>

std::string get_user_details_by_id(const std::vector<std::string>& user_ids, int user_id) {
    if (user_id < 0 || user_id >= static_cast<int>(user_ids.size())) {
        throw std::out_of_range("user_id out of range");
    }
    return user_ids[user_id];
}

int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_failures++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception raised as expected
    } catch (...) {
        // Non-standard exception raised, still counts as raised
    }
}

int main() {
    // Test valid lookups
    expect_equal(get_user_details_by_id({"userA", "userB", "userC"}, 1), "userB", "Test 1");
    expect_equal(get_user_details_by_id({"u0"}, 0), "u0", "Test 2");
    expect_equal(get_user_details_by_id({"x", "y", "z"}, 2), "z", "Test 3");
    expect_equal(get_user_details_by_id({"a", "b"}, 0), "a", "Test 4");

    // Test out of range index (equivalent to Python's IndexError)
    assert_raises([]() { get_user_details_by_id({"userA", "userB", "userC"}, -1); }, "Negative index");
    assert_raises([]() { get_user_details_by_id({"userA", "userB", "userC"}, 3); }, "Index too large");

    // Test type safety (C++ enforces type safety at compile time, so we verify that 
    // invalid types like strings for user_id or non-vector containers cannot be passed)
    // The Python tests check for TypeError at runtime. In C++, these would be compile-time errors.
    // We ensure the function signature strictly enforces std::vector<std::string> and int.
    
    // To simulate the Python test's intent of ensuring bad types don't work, 
    // we confirm the function cannot be invoked with mismatched types.
    // This is inherently guaranteed by C++'s static type system.
    // The following lines would cause a compilation error if uncommented, proving type safety:
    // get_user_details_by_id("not a list", 0); // Error: cannot convert 'const char*' to 'std::vector<std::string>'
    // get_user_details_by_id({"userA", "userB"}, "1"); // Error: cannot convert 'const char*' to 'int'
    
    expect_true(true, "C++ static type checking prevents invalid argument types at compile time");

    if (test_failures > 0) {
        return 1;
    }
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-125_07/main.go
package main

import (
	"errors"
	"fmt"
	"os"
)
// TypeError is returned when the input types are invalid.
type TypeError struct {
	Msg string
}

func (e TypeError) Error() string {
	return e.Msg
}

// IndexError is returned when the user_id is out of the valid range.
type IndexError struct {
	Msg string
}

func (e IndexError) Error() string {
	return e.Msg
}

// getUserDetailsByID returns the entry at position userID from the list userIDs.
// It returns an appropriate error when it cannot complete the task successfully.
func getUserDetailsByID(userIDs []string, userID int) (string, error) {
	if userIDs == nil {
		return "", TypeError{Msg: "user_ids must be a list of strings"}
	}
	if userID < 0 || userID >= len(userIDs) {
		return "", IndexError{Msg: "user_id out of range"}
	}
	return userIDs[userID], nil
}

func main() {
	passed := true

	// assertEqual checks if the function returns the expected string successfully.
	assertEqual := func(userIDs []string, userID int, expected string) {
		result, err := getUserDetailsByID(userIDs, userID)
		if err != nil {
			fmt.Printf("FAIL: getUserDetailsByID(%v, %d) returned unexpected error: %v\n", userIDs, userID, err)
			passed = false
			return
		}
		if result != expected {
			fmt.Printf("FAIL: getUserDetailsByID(%v, %d) = %q; want %q\n", userIDs, userID, result, expected)
			passed = false
		}
	}

	// assertRaises checks if the function returns the expected error type.
	assertRaises := func(userIDs []string, userID int, excType error) {
		_, err := getUserDetailsByID(userIDs, userID)
		if err == nil {
			fmt.Printf("FAIL: getUserDetailsByID(%v, %d) expected an error but got nil\n", userIDs, userID)
			passed = false
			return
		}
		if !errors.As(err, excType) {
			fmt.Printf("FAIL: getUserDetailsByID(%v, %d) returned error type %T; want %T\n", userIDs, userID, err, excType)
			passed = false
		}
	}

	// Valid inputs
	assertEqual([]string{"userA", "userB", "userC"}, 1, "userB")
	assertEqual([]string{"u0"}, 0, "u0")
	assertEqual([]string{"x", "y", "z"}, 2, "z")
	assertEqual([]string{"a", "b"}, 0, "a")

	// Must raise on invalid index
	assertRaises([]string{"userA", "userB", "userC"}, -1, &IndexError{})
	assertRaises([]string{"userA", "userB", "userC"}, 3, &IndexError{})

	// Type errors must raise (nil slice simulates "not a list" in Go type system)
	assertRaises(nil, 0, &TypeError{})

	if !passed {
		os.Exit(1)
	}

	fmt.Println("All tests passed!")
}